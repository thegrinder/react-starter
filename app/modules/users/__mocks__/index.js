const promise = () => Promise.resolve({ data: {} });

export const fetchUser = jest.fn(promise);

export const fetchUsers = jest.fn(promise);
