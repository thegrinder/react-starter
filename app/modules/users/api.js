import axios from '../axios';

export const fetchUsers = (queryParams) =>
  axios.get('/users', { params: queryParams });

export const fetchUser = (id) => axios.get(`/users/${id}`);
