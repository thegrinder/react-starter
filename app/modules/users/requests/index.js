export { fetchUserActions } from './fetchUser';
export { fetchUsersActions } from './fetchUsers';
