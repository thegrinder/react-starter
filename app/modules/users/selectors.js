import { useSelector } from 'react-redux';

// selectors
export const usersStateSelector = (state) => state.users;

export const usersSelector = (state) => usersStateSelector(state).data;

export const userSelector = (state, id) => usersStateSelector(state).data[id];

export const fetchUsersRequestSelector = (state) =>
  usersStateSelector(state).requests.fetchUsers;

export const fetchUserRequestSelector = (state) =>
  usersStateSelector(state).requests.fetchUser;

// hooks
export const useUsersStateSelector = () => useSelector(usersStateSelector);

export const useUsersSelector = () => useSelector(usersSelector);

export const useUserSelector = (id) =>
  useSelector((state) => userSelector(state, id), [id]);

export const useFetchUsersRequestSelector = () =>
  useSelector(fetchUsersRequestSelector);

export const useFetchUserRequestSelector = () =>
  useSelector(fetchUserRequestSelector);
