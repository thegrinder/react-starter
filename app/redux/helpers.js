export const initialRequestState = {
  error: {},
  initialLoad: true,
  loading: false,
};

export const createRequest = (namespace) => {
  const actionTypes = {
    FAILED: `${namespace}/FAILED`,
    FULFILLED: `${namespace}/FULFILLED`,
    LOADING: `${namespace}/LOADING`,
    RESET: `${namespace}/RESET`,
    SUCCEEDED: `${namespace}/SUCCEEDED`,
    TRIGGER: `${namespace}/TRIGGER`,
  };

  const actionCreators = {
    failed: error => ({ error, type: actionTypes.FAILED }),
    fulfilled: () => ({ type: actionTypes.FULFILLED }),
    loading: () => ({ type: actionTypes.LOADING }),
    reset: () => ({ type: actionTypes.RESET }),
  };

  const reducer = (state = initialRequestState, action) => {
    switch (action.type) {
      case actionTypes.LOADING:
        return {
          ...state,
          error: {},
          loading: true,
        };
      case actionTypes.FAILED:
        return {
          ...state,
          error: action.error,
        };
      case actionTypes.FULFILLED:
        return {
          ...state,
          initialLoad: false,
          loading: false,
        };
      case actionTypes.RESET:
        return initialRequestState;
      default:
        return state;
    }
  };

  return {
    actionCreators,
    actionTypes,
    reducer,
  };
};
