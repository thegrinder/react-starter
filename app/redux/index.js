export * from './selectors';
export { fetchReposActions, fetchReposActionTypes } from './repos/fetchRepos';
