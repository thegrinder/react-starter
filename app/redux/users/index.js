export * from './requests';
export * from './selectors';
