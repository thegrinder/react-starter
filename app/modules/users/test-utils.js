import { initialRequestState } from '../helpers';
import { initialUsersState } from './reducer';

export const createUser = (user = {}) => ({
  id: 'id',
  name: 'name',
  username: 'username',
  email: 'email@email.com',
  phone: '132456776',
  ...user,
});

export const createRequestState = (requestState = {}) => ({
  ...initialRequestState,
  ...requestState,
});

export const createUsersState = ({
  requestKey = '',
  requestState = {},
  data = {},
} = {}) => ({
  ...initialUsersState,
  data: {
    ...initialUsersState.data,
    ...data,
  },
  requests: {
    ...initialUsersState.requests,
    ...(requestKey && {
      [requestKey]: {
        ...initialUsersState.requests[requestKey],
        ...requestState,
      },
    }),
  },
});

export const createState = args => ({
  users: createUsersState(args),
});
