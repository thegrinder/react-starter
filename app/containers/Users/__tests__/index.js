import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';

import {
  render,
  fireEvent,
  act,
  createStore,
} from '../../../helpers/test-utils';
import { fetchUsers } from '../../../redux/users';
import { createState, createRequestState } from '../../../redux/users/test-uitls';
import { Users } from '..';


const renderComponent = store => render(
  <MemoryRouter>
    <Provider store={store}>
      <Users />
    </Provider>
  </MemoryRouter>,
);

describe('<Users />', () => {
  it('should render correctly', () => {
    const state = createState();
    const store = createStore(state);
    const { container: { firstChild } } = renderComponent(store);
    expect(firstChild).toBeDefined();
    expect(firstChild).toMatchSnapshot();
  });

  it('should search users with a given name on submit button click', () => {
    const state = createState();
    const store = createStore(state);
    const { getByText, getByPlaceholderText } = renderComponent(store);
    const name = 'text';
    act(() => {
      fireEvent.change(getByPlaceholderText('Search'), { target: { value: name } });
    });
    act(() => {
      fireEvent.click(getByText('Search'));
    });
    expect(store.getActions()).toContainEqual(fetchUsers({ name }));
  });

  it('should render the spinner when the request is being made', () => {
    const state = createState({
      requestKey: 'fetchUsers',
      requestState: createRequestState({ loading: true }),
    });
    const store = createStore(state);
    const { getByLabelText } = renderComponent(store);
    expect(getByLabelText('loading...')).toBeInTheDocument();
  });
});
