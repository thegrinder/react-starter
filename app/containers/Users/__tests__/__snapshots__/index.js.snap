// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Users /> should render correctly 1`] = `
<div
  class="pv-6 sc-bdVaJa hGfkPm"
>
  <div
    class="pa-8 sc-htpNat fEfKmN"
  >
    <form
      class="flex flex-row items-start -mh-4"
    >
      <div
        class="col ph-4"
      >
        <div
          class="mb-2"
        >
          
          <input
            class="StyledInput-ok3pz6-0 buStls"
            id="search"
            placeholder="Search"
            value=""
          />
          <div
            class="h-4"
          />
        </div>
      </div>
      <div
        class="col-auto ph-4"
      >
        <button
          class="StyledButton-k0ad56-0 hzwlwW"
          type="submit"
        >
          <span
            class="ChildrenWrapper-sc-30uoyc-0 cdOdhF"
          >
            Search
          </span>
        </button>
      </div>
    </form>
    <div>
      <h5
        class="StyledHeading-fbinjh-0 geivid"
        color="default"
      >
        Results:
      </h5>
      <ul />
    </div>
  </div>
</div>
`;
